### PR TITLE
Record match & scoring data

### PR DESCRIPTION
**Pull request by PeterJCLaw**
This has the competition mode runs of the simulator output a JSON file which is compatible with the [Proton](https://github.com/PeterJCLaw/proton) specification used by SRComp for scoring.

This will also allow us to more easily move the robot logs to suitable directories for storage, though that is not addressed here.

This contributes towards #18, #210 and #189.

Suggest review by commit.

Note: this may end up conflicting with #240, though hopefully not too badly.

See https://github.com/srobo/sr2021-comp/pull/1 for the scorer implementation.

This is part 1 of 3 towards match automation:
- Record match & scoring data (this PR #244)
- Add recording controls (#254)
- Automate running of individual matches (#256)

### CODE REVIEW

File: script/prepare-comp-match [hunk 1]
```diff
@@ -14,8 +14,7 @@ REPO_ROOT = Path(__file__).resolve().parent.parent
 sys.path.insert(1, str(REPO_ROOT / 'modules'))
 
 import controller_utils  # isort:skip
-
-NUM_ZONES = 2
+from controller_utils import NUM_ZONES  # isort:skip
```
Comment (RealOrangeOne) [inline script/prepare-comp-match:17]: This seems unneessary, can you not just replace the uses of `NUM_ZONES` with `controller_utils.NUM_ZONES`?

Comment (PeterJCLaw) [reply script/prepare-comp-match:17]: For sure we could, this felt like it left that code slightly clearer though (and avoided some slightly odd wrapping).

Comment (PeterJCLaw) [reply script/prepare-comp-match]: 0dfc03781cc75adb159aab5bceb513c618ce86ac

**Review by WillB97 — APPROVED** [review]
Once #256 and #254 have been merged.

**Comment by PeterJCLaw** [comment]
> Once #256 and #254 have been merged.

Why couple those changes to this? This PR doesn't depend on those (somewhat deliberately so); rather they depend upon this PR. While these changes were designed with those subsequent changes in mind, they were also designed to be useful and sufficient on their own.

**Comment by WillB97** [comment]
I was incorrect, but #256 is needed since the excess logging I/O which was fixed there appears to effect this PR as well.

**Comment by PeterJCLaw** [comment]
> I was incorrect, but #256 is needed since the excess logging I/O which was fixed there appears to effect this PR as well.

Ah, I see. I'd rather address that by pulling the I/O improvements here and then merging. Coupling this PR to those just for that fix doesn't seem like a great solution to me. I'll do some branch juggling shortly to arrange that.

**Pull request merged**